File: ocaml/src/vectorize/block_builder.ml
open Llvm

module ControlDependence = struct
  type t
  type condition_and
  type condition_or
end

type block_builder = {
  ctx : llcontext;
  f : llvalue; 
  emit_condition : llvalue -> llvalue;
  mutable active_conds : (ControlDependence.t option * llbasicblock) list;
  mutable semi_active_conds : (ControlDependence.t * ControlDependence.t list) list;
  mutable dummy_counter : int;
}

let create_block_builder entry_bb emit_condition =
  let ctx = basic_block_context entry_bb in
  let f = block_parent entry_bb in
  {
    ctx;
    f;
    emit_condition;
    active_conds = [(None, entry_bb)];
    semi_active_conds = [];
    dummy_counter = 0;
  }

let create_block bb = append_block (basic_block_context bb) "" (block_parent bb)

module ConditionEmitter = struct
  type t = {
    builder : llbuilder;
    common : ControlDependence.t option;
    emit_condition : llvalue -> llvalue;
    mutable emitted : (ControlDependence.t, llvalue) Hashtbl.t;
  }

  let create bb common emit_condition =
    {
      builder = builder_at_end (basic_block_context bb) bb;
      common;
      emit_condition;
      emitted = Hashtbl.create 16;
    }

  let rec emit ce c =
    match c with
    | None -> const_int (i1_type ce.builder.context) 1
    | Some c when c = ce.common -> const_int (i1_type ce.builder.context) 1
    | Some (ControlDependence.ConditionAnd and_c) -> emit_and ce and_c
    | Some (ControlDependence.ConditionOr or_c) -> emit_or ce or_c

  let emit_and ce and_c =
    match Hashtbl.find_opt ce.emitted and_c with
    | Some v -> v
    | None ->
        let v =
          build_and
            (emit ce and_c.parent)
            (if and_c.is_true then
               ce.emit_condition and_c.cond
             else
               build_not (ce.emit_condition and_c.cond) "" ce.builder)
            "" ce.builder
        in
        Hashtbl.add ce.emitted and_c v;
        v

  let emit_or ce or_c =
    match Hashtbl.find_opt ce.emitted or_c with
    | Some v -> v
    | None ->
        let v = emit_disjunction ce or_c.conds in
        Hashtbl.add ce.emitted or_c v;
        v

  let emit_disjunction ce conds =
    let values = List.map (emit ce) conds in
    build_or (Array.of_list values) "" ce.builder
end

let get_block_for bb c =
  let rec find_active_conds c conds =
    let visited = Hashtbl.create 16 in
    let rec dfs c2 =
      if not (Hashtbl.mem visited c2) then begin
        Hashtbl.add visited c2 ();
        if List.mem_assoc (Some c2) bb.active_conds then
          Hashtbl.add conds c2 ()
        else
          match List.assoc_opt c2 bb.semi_active_conds with
          | Some children -> List.iter dfs children
          | None -> ()
      end
    in
    dfs c;
    Hashtbl.to_seq_keys conds |> List.of_seq
  in

  match List.assoc_opt (Some c) bb.active_conds with
  | Some block -> block
  | None ->
      if List.mem_assoc c bb.semi_active_conds then
        let conds = find_active_conds c (Hashtbl.create 16) in
        let new_bb = create_block bb.f in
        List.iter
          (fun c2 ->
            let old_bb = List.assoc (Some c2) bb.active_conds in
            build_br new_bb (builder_at_end bb.ctx old_bb);
            bb.active_conds <- List.remove_assoc (Some c2) bb.active_conds)
          conds;
        bb.active_conds <- (Some c, new_bb) :: bb.active_conds;
        new_bb
      else
        match c with
        | ControlDependence.ConditionAnd and_c ->
            let if_true = create_block bb.f in
            let if_false = create_block bb.f in
            let parent_bb = get_block_for bb and_c.parent in
            let cond = bb.emit_condition and_c.cond in
            build_cond_br cond if_true if_false (builder_at_end bb.ctx parent_bb);
            let result_bb = if and_c.is_true then if_true else if_false in
            bb.active_conds <- List.remove_assoc (Some and_c.parent) bb.active_conds;
            bb.semi_active_conds <- (and_c.parent, [and_c; and_c.complement]) :: bb.semi_active_conds;
            bb.active_conds <- (Some and_c, result_bb) :: bb.active_conds;
            bb.active_conds <- (Some and_c.complement, if and_c.is_true then if_false else if_true) :: bb.active_conds;
            result_bb
        | ControlDependence.ConditionOr or_c ->
            let common_c = or_c.greatest_common_cond in
            let conds =
              if not (List.mem_assoc common_c bb.semi_active_conds) then begin
                ignore (get_block_for bb common_c);
                [common_c]
              end else
                find_active_conds common_c (Hashtbl.create 16)
            in
            let new_bb = create_block bb.f in
            let aux_bb = create_block bb.f in
            let conds_to_join = Hashtbl.create 16 in
            List.iter (fun c -> Hashtbl.add conds_to_join c ()) or_c.conds;
            let joined = Hashtbl.create 16 in
            List.iter
              (fun c2 ->
                let old_bb = List.assoc (Some c2) bb.active_conds in
                if Hashtbl.mem conds_to_join c2 then begin
                  build_br new_bb (builder_at_end bb.ctx old_bb);
                  Hashtbl.add joined c2 ()
                end else
                  build_br aux_bb (builder_at_end bb.ctx old_bb);
                bb.active_conds <- List.remove_assoc (Some c2) bb.active_conds)
              conds;
            let unjoined_conds =
              List.filter (fun c -> not (Hashtbl.mem joined c)) or_c.conds
            in
            let drain_bb = create_block bb.f in
            if unjoined_conds = [] then
              build_br drain_bb (builder_at_end bb.ctx aux_bb)
            else
              let ce = ConditionEmitter.create aux_bb (Some common_c) bb.emit_condition in
              let cond = ConditionEmitter.emit_disjunction ce unjoined_conds in
              build_cond_br cond new_bb drain_bb (builder_at_end bb.ctx aux_bb);
            let dummy_c = get_dummy_condition bb in
            bb.active_conds <- (Some c, new_bb) :: bb.active_conds;
            bb.active_conds <- (Some dummy_c, drain_bb) :: bb.active_conds;
            bb.semi_active_conds <- (common_c, [c; dummy_c]) :: bb.semi_active_conds;
            new_bb

let get_dummy_condition bb =
  bb.dummy_counter <- bb.dummy_counter + 1;
  Obj.magic bb.dummy_counter

let set_block_for_condition bb block c =
  assert (List.mem_assoc (Some c) bb.active_conds);
  bb.active_conds <- (Some c, block) :: List.remove_assoc (Some c) bb.active_conds

File: c++/src/vectorize/BlockBuilder.cpp
#include "BlockBuilder.h"
#include "ControlDependence.h"
#include "llvm/IR/IRBuilder.h"

using namespace llvm;

BlockBuilder::BlockBuilder(
    BasicBlock *EntryBB,
    std::function<llvm::Value *(llvm::Value *)> EmitCondition)
    : Ctx(EntryBB->getContext()), F(EntryBB->getParent()),
      EmitCondition(EmitCondition), ActiveConds({{nullptr, EntryBB}}) {}

BasicBlock *BlockBuilder::createBlock() {
  return BasicBlock::Create(Ctx, "", F);
}

namespace {

class ConditionEmitter {
  IRBuilder<> IRB;
  const ControlCondition *Common;
  std::function<llvm::Value *(llvm::Value *)> EmitCondition;
  DenseMap<const ControlCondition *, Value *> Emitted;

  Value *emitAnd(const ConditionAnd *And) {
    if (auto *V = Emitted.lookup(And))
      return V;
    auto *V =
        IRB.CreateAnd(emit(And->Parent),
                      And->IsTrue ? EmitCondition(And->Cond)
                                  : IRB.CreateNot(EmitCondition(And->Cond)));
    return Emitted[And] = V;
  }

  Value *emitOr(const ConditionOr *Or) {
    if (auto *V = Emitted.lookup(Or))
      return V;
    return Emitted[Or] = emitDisjunction(Or->Conds);
  }

public:
  ConditionEmitter(BasicBlock *BB, const ControlCondition *Common,
                   decltype(EmitCondition) EmitCondition)
      : IRB(BB), Common(Common), EmitCondition(EmitCondition) {}
  Value *emit(const ControlCondition *C) {
    if (C == Common || !C)
      return IRB.getTrue();
    if (auto *And = dyn_cast<ConditionAnd>(C))
      return emitAnd(And);
    return emitOr(cast<ConditionOr>(C));
  }
  Value *emitDisjunction(ArrayRef<const ControlCondition *> Conds) {
    SmallVector<Value *, 8> Values;
    for (auto *C : Conds)
      Values.push_back(emit(C));
    return IRB.CreateOr(Values);
  }
};

} // namespace

BasicBlock *BlockBuilder::getBlockFor(const ControlCondition *C) {
  if (auto *BB = ActiveConds.lookup(C))
    return BB;

  // Get active conditions that use C and
  // unmark all intermediate semi-active conditions.
  auto GetActiveConds = [&](const ControlCondition *C,
                            SmallPtrSetImpl<const ControlCondition *> &Conds) {
    SmallPtrSet<const ControlCondition *, 4> Visited;
    SmallVector<const ControlCondition *> Worklist{C};
    assert(SemiActiveConds.count(C));
    while (!Worklist.empty()) {
      auto *C2 = Worklist.pop_back_val();
      if (!Visited.insert(C2).second)
        continue;

      if (ActiveConds.count(C2)) {
        Conds.insert(C2);
        continue;
      }

      auto It = SemiActiveConds.find(C2);
      assert(It != SemiActiveConds.end());
      Worklist.append(It->second.begin(), It->second.end());
      SemiActiveConds.erase(It);
    }
  };

  // If C is a semi-active condition,
  // join all of the blocks using C to the block for C
  if (SemiActiveConds.count(C)) {
    SmallPtrSet<const ControlCondition *, 4> Conds;
    GetActiveConds(C, Conds);
    auto *BB = createBlock();
    for (auto *C2 : Conds) {
      auto It = ActiveConds.find(C2);
      assert(It != ActiveConds.end());
      BranchInst::Create(BB, It->second);
      ActiveConds.erase(It);
    }
    ActiveConds[C] = BB;
    return BB;
  }

  if (auto *And = dyn_cast<ConditionAnd>(C)) {
    auto *IfTrue = createBlock();
    auto *IfFalse = createBlock();
    BranchInst::Create(IfTrue, IfFalse, EmitCondition(And->Cond),
                       getBlockFor(And->Parent));
    auto *BB = And->IsTrue ? IfTrue : IfFalse;

    assert(ActiveConds.count(And->Parent));
    ActiveConds.erase(And->Parent);
    SemiActiveConds[And->Parent].assign({And, And->Complement});
    ActiveConds[And] = BB;
    ActiveConds[And->Complement] = And->IsTrue ? IfFalse : IfTrue;
    return BB;
  }

  auto *Or = cast<ConditionOr>(C);
  SmallPtrSet<const ControlCondition *, 4> Conds;
  auto *CommonC = Or->GreatestCommonCond;
  if (!SemiActiveConds.count(CommonC)) {
    (void)getBlockFor(CommonC);
    Conds.insert(CommonC);
  } else {
    GetActiveConds(CommonC, Conds);
  }

  auto *BB = createBlock();
  auto *AuxBB = createBlock();
  SmallPtrSet<const ControlCondition *, 4> CondsToJoin, Joined;
  CondsToJoin.insert(Or->Conds.begin(), Or->Conds.end());
  for (auto *C2 : Conds) {
    auto It = ActiveConds.find(C2);
    assert(It != ActiveConds.end());
    auto *BB2 = It->second;

    if (CondsToJoin.count(C2)) {
      BranchInst::Create(BB, BB2);
      Joined.insert(C2);
    } else {
      BranchInst::Create(AuxBB, BB2);
    }
    ActiveConds.erase(It);
  }

  SmallVector<const ControlCondition *, 4> UnjoinedConds;
  for (auto *C : Or->Conds)
    if (!Joined.count(C))
      UnjoinedConds.push_back(C);

  // Branch conditionally from AuxBB to BB
  auto *DrainBB = createBlock();
  if (UnjoinedConds.empty())
    BranchInst::Create(DrainBB, AuxBB);
  else {
    ConditionEmitter CE(AuxBB, CommonC, EmitCondition);
    BranchInst::Create(BB, DrainBB,
                       CE.emitDisjunction(UnjoinedConds), AuxBB);
  }

  auto *DummyC = getDummyCondition();
  ActiveConds[C] = BB;
  ActiveConds[DummyC] = DrainBB;
  SemiActiveConds[CommonC] = {C, DummyC};
  assert(!ActiveConds.lookup(CommonC));
  return BB;
}

void BlockBuilder::setBlockForCondition(llvm::BasicBlock *BB,
                                        const ControlCondition *C) {
  assert(ActiveConds.count(C) && "can only set block for active condition");
  ActiveConds[C] = BB;
}

const ControlCondition *BlockBuilder::getDummyCondition() {
  return reinterpret_cast<const ControlCondition *>(++DummyCounter);
}